optimize docs for event system (#1903)
cocos-creator/creator-docs: en/engine/event/event-node.md

# Node Event System

> As mentioned in the previous document, the `input` object supports the global [input event system](event-input.md), and the `Node` also implements the event listening interface of the `EventTarget`. On this basis, we provide a basic node-related system events. This document will introduce how to use these events.

Reviewer: ```suggestion
As mentioned in the previous document, the `input` object supports the global [input event system](event-input.md), and the `Node` also implements the event listening interface of the `EventTarget`. On this basis, the basic node-related system events are provided. This document will introduce how to use these events.
```

__Cocos Creator__ supports four types of system events: __mouse__, __touch__, __keyboard__, __device motion__. They are called [Global Input Events](event-input.md). The usage of touch and mouse events dispatched by `Node` is discussed in this document. The usage is as follows:

```ts
node.on(Node.EventType.MOUSE_DOWN, (event) => {
  console.log('Mouse down');
}, this);
```

> > __Note__: We no longer recommend directly using the event name string to register event listeners. Like the example above, please do not use `node.on('mouse-down', callback, target)` to register event listeners.

Reviewer: ```suggestion
> __Note__: It is no longer recommend directly using the event name string to register event listeners. Like the example above, please do not use `node.on('mouse-down', callback, target)` to register event listeners.
```

Author: I change in this way https://github.com/cocos-creator/creator-docs/pull/1903/commits/1f4a467d00b8b7e616c55458509fc0425af16d56

> The touch event listener on the node depends on the `UITransform` component, which is only applicable to 2D UI nodes. If you need to implement the touch detection of 3D objects, you can refer to the document [Touch Detection of 3D Objects](event-input.md#Touch%20detection%20for%203D%20objects)

Reviewer: ```suggestion
The touch event listener on the node depends on the `UITransform` component, which is only applicable to 2D UI nodes. To implement the touch detection of 3D objects, refer to the [Touch Detection of 3D Objects](event-input.md#Touch%20detection%20for%203D%20objects) documentation.
```

| Enumerated Object Definition  | Timing of Event Triggering                                                               |
|:--------------------------------|:-----------------------------------------------------------------------------------------|
| __Node.EventType.TOUCH_START__  | When one or more touch points are placed in the target node region                       |
| __Node.EventType.TOUCH_MOVE__   | When one or more touch points are moved along the screen                                 |
| __Node.EventType.TOUCH_END__    | When one or more touch points are removed from the screen in the target node region      |
| __Node.EventType.TOUCH_CANCEL__ | When one or more touch points are removed from the screen outside the target node region |

The important APIs of a touch event (`Event.EventTouch`) are described in the [Mouse Events API](event-api.md#Touch-Event-API) (`Event` standard event API excluded):

> __Note__: touch events support multi-touch, each touch spot will send one event to the event listener.

## Node Event Dispatching

> We support `dispatchEvent` interface on `Node`. Events dispatched by this interface, would enter the event delivery stage. The event dispatching system of Cocos Creator is based on the implementation of [event bubbling and capture on Web standard](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_bubbling_and_capture). After the event is dispatched, it will go through the following three stages:

Reviewer: ```suggestion
The `dispatchEvent` interface is supported on `Node`. Events dispatched by this interface, would enter the event delivery stage. The event dispatching system of Cocos Creator is based on the implementation of [event bubbling and capture on Web standard](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_bubbling_and_capture). After the event is dispatched, it will go through the following three stages:
```

- __capturing phase__: The event is passed from the scene root node to the child nodes step by step, until it reaches the target node or the event propagation is stopped in the event callback
- __target phase__: The event is triggered on the target node
- __bubbling phase__: The event is bubbled from the target node to the parent node level by level, until the root node is reached or the event propagation is stopped in the event callback

> When we call `node.dispatchEvent()`, it means that `node` is the target node mentioned above. In the process of event delivery, we can stop the event propagation by calling `event.propagationStopped = true`.

Reviewer: ```suggestion
When calling `node.dispatchEvent()`, it means that `node` is the target node mentioned above. In the process of event delivery, to stop the event propagation by call `event.propagationStopped = true`.
```

Author: I change this way https://github.com/cocos-creator/creator-docs/pull/1903/commits/0a6ac9d9c7777a7d03720b2cb905a237de48cf91 https://github.com/cocos-creator/creator-docs/pull/1903/commits/0c3fe09cb338f552782865ac09f2bb447b29d885

> In v3.0, we removed the `Event.EventCustom` class. To dispatch custom events, a custom event class that inherits from the `Event` class needs to be implemented first. For example:

Reviewer: ```suggestion
In v3.0, the `Event.EventCustom` class was removed. To dispatch custom events, a custom event class that inherits from the `Event` class needs to be implemented first. For example:
```

| __target__                      | Node     | The original target that received the event.                                                                                                     |
| __currentTarget__               | Node     | The current object that received the event. The current target of the event during the bubbling phase may be different from the original target. |
| __getType__                     | Function | Get the event type.                                                                                                                              |
| __propagationStopped__          | Boolean  | Whether or not stop the bubbling phase. The parent node of the current target no longer receives the corresponding event.                        |
| __propagationImmediateStopped__ | Boolean  | Whether or not stop passing the current event immediately. The current target no longer receives the event either.                               |

Please refer to the `Event` and API files of its child category for a complete API list.

## Touch event propagation

> As mentioned above, touch events registered on the `Node` are dispatched internally by the engine through the `dispatchEvent` interface. Below we will introduce the propagation sequence of touch events in the **target** and **bubbling** phases.

Reviewer: ```suggestion
As mentioned above, touch events registered on the `Node` are dispatched internally by the engine through the `dispatchEvent` interface. Below introduces the propagation sequence of touch events in the **target** and **bubbling** phases.
```

### Touch event bubbling

Touch events support the event bubbling on the node tree, take the pictures below as an example:

![propagation](propagation.png)

In the scene shown in the picture, suppose node A has a child node B which has a child node C. The developer sets the touch event listeners for all these three nodes (each node has a touch event listener in examples below by default).

When the mouse or finger was applied in the node C region, the event will be triggered at node C first and the node C listener will receive the event (this is the target phase). Then the node C will pass this event to its parent node, so the node B listener will receive this event. Similarly the node B will also pass the event to its parent node A. This is a basic event bubbling phase. 

> > __Note__: It needs to be emphasized that there is no hit test in parent nodes in the bubbling phase, which means that the node A and B can receive touch events even though the touch location is out of their node region.

Reviewer: ```suggestion
> __Note__: it needs to be emphasized that there is no hit test in parent nodes in the bubbling phase, which means that nodes A and B can receive touch events even though the touch location is out of their node region.
```

> The bubbling phase of touch events is no different than the general events. Calling `event.propagationStopped = true` can force to stop the bubbling phase.

Reviewer: ```suggestion
The bubbling phase of touch events is no different than the general events. Calling `event.propagationStopped = true` can force stopping the bubbling phase.
```

![multi-canvas](multi-canvas.png)

### Register touch or mouse events in the capturing phase

Sometimes, it is necessary to dispatch touch or mouse events to parent node event listeners before dispatching to any child nodes beneath it in hierarchy, like the design of __ScrollView__ component.

Event bubbling cannot meet all demands. When this happens, register the parent node event listeners in the capturing phase. To achieve this goal, passing in `true` as the `useCapture` parameter (a fourth parameter) when registering touch or mouse event on the node. For example:

```ts
this.node.on(Node.EventType.TOUCH_START, this.onTouchStartCallback, this, true);
```

When the node fires the `touch-start` event, the `touch-start` event will be first dispatched to all the parent node event listeners registered in the capturing phase, then dispatched to the node itself, and finally comes the event bubbling phase.

Only touch or mouse events can be registered in the capturing phase, while the other events can't be.

### Event Interception

> Normal events are dispensed as described above. However, if the node has components such as __Button__, __Toggle__ or __BlockInputEvents__ on it, it will stop event bubbling. Look at the picture below. There are two buttons, priority 1 for Canvas0 and priority 2 for Canvas1. If you click on the intersection of the two buttons, which is the blue area in the picture, it appears that button priority 2 received the contact event successfully, while button priority 1 did not. That's because according to the event reception rules above, button priority 2 receives contact events first and intercepts them (`event.propagationStopped = true`) to prevent event penetration. If the node is a non-button node, events can also be intercepted by adding the __BlockInputEvents__ component to prevent penetration.

Reviewer: ```suggestion
Normal events are dispensed as described above. However, if the node has components such as __Button__, __Toggle__ or __BlockInputEvents__ on it, it will stop event bubbling. Look at the picture below. There are two buttons, priority 1 for Canvas0 and priority 2 for Canvas1. Click on the intersection of the two buttons, which is the blue area in the picture, it appears that button priority 2 received the contact event successfully, while button priority 1 did not. That's because according to the event reception rules above, button priority 2 receives contact events first and intercepts them (`event.propagationStopped = true`) to prevent event penetration. If the node is a non-button node, events can also be intercepted by adding the __BlockInputEvents__ component to prevent penetration.
```

![events-block](events-block.png)

## Example for touch events

Using the example below, summarizing touch events is easy. There are four nodes A, B, C and D in the picture above, where A and B are brother nodes. The specific hierarchical relationship should be like this:

![example](example.png)

1. If one touch is applied in the overlapping area between A and B, now B won't receive the touch event, so that propagating order of the touch event should be __A -> C -> D__.
2. If the touch location is in node B (the visible green area), the order should be __B -> C -> D__.
3. If the touch location is in node C, the order should be __C -> D__.

> 4. As a precondition to the second case, we register touch events on C D node in the capturing phase, then the order should be __D -> C -> B__.

Reviewer: ```suggestion
4. As a precondition to the second case, register touch events on C D node in the capturing phase, then the order should be __D -> C -> B__.
```
